### ISSUE TITLE
msg.Len() disagrees with Pack()

### ISSUE THREAD

**Issue by gpaul** [miekg/dns#453]
Here's a test that fails:

```

func TestMsgCompressLength2(t *testing.T) {
	msg := new(Msg)
	msg.Compress = true
	msg.SetQuestion(Fqdn("_foo._tcp.bar."), TypeANY)
	msg.Answer = append(msg.Answer, &SRV{Hdr: RR_Header{Name: "_foo._tcp.bar.", Rrtype: 0x21, Class: 0x1, Ttl: 0x3c}, Port: 0x4c57, Target: "foo.bar."})
	msg.Answer = append(msg.Answer, &SRV{Hdr: RR_Header{Name: "_foo._tcp.bar.", Rrtype: 0x21, Class: 0x1, Ttl: 0x3c}, Port: 0x3307, Target: "foo.bar."})
	msg.Answer = append(msg.Answer, &SRV{Hdr: RR_Header{Name: "_foo._tcp.bar.", Rrtype: 0x21, Class: 0x1, Ttl: 0x3c}, Port: 0x35b0, Target: "foo.bar."})
	msg.Answer = append(msg.Answer, &SRV{Hdr: RR_Header{Name: "_foo._tcp.bar.", Rrtype: 0x21, Class: 0x1, Ttl: 0x3c}, Port: 0x5a7a, Target: "foo.bar."})
	msg.Answer = append(msg.Answer, &SRV{Hdr: RR_Header{Name: "_foo._tcp.bar.", Rrtype: 0x21, Class: 0x1, Ttl: 0x3c}, Port: 0x211a, Target: "foo.bar."})
	msg.Extra = append(msg.Extra, &A{Hdr: RR_Header{Name: "foo.bar.", Rrtype: 0x1, Class: 0x1, Ttl: 0x3c}, A: net.IP{0xac, 0x11, 0x0, 0x3}})
	msg.Extra = append(msg.Extra, &A{Hdr: RR_Header{Name: "foo.bar.", Rrtype: 0x1, Class: 0x1, Ttl: 0x3c}, A: net.IP{0xac, 0x11, 0x0, 0x3}})
	msg.Extra = append(msg.Extra, &A{Hdr: RR_Header{Name: "foo.bar.", Rrtype: 0x1, Class: 0x1, Ttl: 0x3c}, A: net.IP{0xac, 0x11, 0x0, 0x3}})
	msg.Extra = append(msg.Extra, &A{Hdr: RR_Header{Name: "foo.bar.", Rrtype: 0x1, Class: 0x1, Ttl: 0x3c}, A: net.IP{0xac, 0x11, 0x0, 0x3}})
	msg.Extra = append(msg.Extra, &A{Hdr: RR_Header{Name: "foo.bar.", Rrtype: 0x1, Class: 0x1, Ttl: 0x3c}, A: net.IP{0xac, 0x11, 0x0, 0x3}})
	predicted := msg.Len()
	buf, err := msg.Pack()
	if err != nil {
		t.Error(err)
	}
	if predicted != len(buf) {
		t.Errorf("predicted compressed length is wrong: predicted %s (len=%d) %d, actual %d",
			msg.Question[0].Name, len(msg.Answer), predicted, len(buf))
	}
}
```

Results in 

```
=== RUN   TestMsgCompressLength2
--- FAIL: TestMsgCompressLength2 (0.00s)
	dns_test.go:215: predicted compressed length is wrong: predicted _foo._tcp.bar. (len=5) 246, actual 250

```

**Comment by miekg** [comment]
Thanks for the test case.

This is tricky to get right as `Len` and `Pack` use 2 diff. algorithms, i.e. you don't want to Pack for every Len you do, and then throw away the buffer.

**Comment by gpaul** [comment]
Totally, this is why I posted a test case and didn't submit a PR :)

**Comment by gpaul** [comment]
Do you think you'll get around to this in the near future? If not I could have a look.

**Comment by miekg** [comment]
This is the problem: https://github.com/miekg/dns/blob/master/msg.go#L994

Didn't want to use reflection at the time, but now that we "go gen" the packs and unpacks this needs to be generated as well. Although looking through the list, SRV is in there

**Comment by miekg** [comment]
Can you make your example smaller?

**Comment by miekg** [comment]
~~~
func TestMsgCompressLength2(t *testing.T) {
	msg := new(Msg)
	msg.Compress = true
	msg.SetQuestion(Fqdn("_foo._tcp.bar."), TypeANY)
	msg.Answer = append(msg.Answer, &SRV{Hdr: RR_Header{Name: "_foo._tcp.bar.", Rrtype: 0x21, Class: 0x1, Ttl: 0x3c}, Port: 0x4c57, Target: "foo.bar."})
	msg.Extra = append(msg.Extra, &A{Hdr: RR_Header{Name: "foo.bar.", Rrtype: 0x1, Class: 0x1, Ttl: 0x3c}, A: net.IP{0xac, 0x11, 0x0, 0x3}})
	predicted := msg.Len()
	buf, err := msg.Pack()
	if err != nil {
		t.Error(err)
	}
	if predicted != len(buf) {
		t.Errorf("predicted compressed length is wrong: predicted %s (len=%d) %d, actual %d",
			msg.Question[0].Name, len(msg.Answer), predicted, len(buf))
	}
}
~~~
shows the problem as well.

**Comment by miekg** [comment]
`compressionLenSearchType` does *not* have SRV; i.e. yes we should `gen` these functions.

**Comment by miekg** [comment]
This is prolly also: https://github.com/skynetservices/skydns/issues/313

**Comment by miekg** [comment]
https://github.com/miekg/dns/pull/454 does *not* fix it, but is still good to have.

**Comment by miekg** [comment]
Updated that PR which fixes the issue. Does that work/help you?

**Comment by gpaul** [comment]
I'll have a look asap - thanks for the fast turn-around

**Comment by gpaul** [comment]
I confirm that this PR fixes the issue for me.

**Comment by miekg** [comment]
Fixed by #454
